File: Practice/add_two_numbers.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        
        ListNode* curr = nullptr;
        ListNode* prev = nullptr;
        ListNode* result = nullptr;

        ListNode* n1 = l1;
        ListNode* n2 = l2;

        int ones = 0;
        int carry = 0;
        int sum = 0;

        while ( !(n1 == nullptr || n2 == nullptr))
        {
            if (n1 != nullptr && n2 != nullptr)
            {
                int k1 = n1->val;
                int k2 = n2->val;

                sum = k1 + k2 + carry;
                ones = sum % 10;
                carry = sum / 10;

                curr = new ListNode(ones);
                if (prev == nullptr)
                    prev = curr;
                else
                    prev->next = curr;

                n1 = n1->next;
                n2 = n2->next;

                printf("%d", ones);

            }
            else if(n1 == nullptr && n2 != nullptr)
            {
                sum = n2->val + carry;
                ones = sum % 10;
                carry = sum / 10;

                curr = new ListNode(ones);
                if (prev == nullptr)
                    prev = curr;
                else
                    prev->next = curr;

                n1 = n1->next;
                n2 = n2->next;
            }
            else if(n2 == nullptr && n1 != nullptr)
            {
                sum = n1->val + carry;
                ones = sum % 10;
                carry = sum / 10;

                curr = new ListNode(ones);
                if (prev == nullptr)
                    prev = curr;
                else
                    prev->next = curr;

                n1 = n1->next;
                n2 = n2->next;
            }
            else if(n1 == nullptr && n2 == nullptr)
            {
                break;
            }

            if (result == nullptr && curr != nullptr)
            {
                result = curr;
            }
        }

        return result;
    }
};

void printNum(std::string name, ListNode* num)
{
    ListNode* digit = num;
    cout << name << ": ";
    while(digit != nullptr)
    {
        printf("%d", digit->val);
        digit = digit->next;
    }
    printf("\n");
}

int main()
{
    ListNode* l1 = new ListNode(2);
    ListNode* l2 = new ListNode(4);
    ListNode* l3 = new ListNode(3);

    ListNode* l4 = new ListNode(5);
    ListNode* l5 = new ListNode(6);
    ListNode* l6 = new ListNode(4);

    l1->next = l2;
    l2->next = l3;

    l4->next = l5;
    l5->next = l6;

    printNum("N1", l1);
    printNum("N2", l4);

    Solution s;
    ListNode* result = s.addTwoNumbers(l1, l4);

    printNum("Result", result);

    return 0;
}

File: Practice/Medium/add_two_numbers.py

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:

        result = None
        r1 = None

        p_carry = 0

        n1 = l1
        n2 = l2
        while n1 != None or n2 != None:

            if n1 == None:
                carry, ones = add(0, n2.val, p_carry)
                n2 = n2.next
                p_carry = 0
            elif n2 == None:
                carry, ones = add(n1.val, 0, p_carry)
                n1 = n1.next
                p_carry = 0
            else:
                carry, ones = add(n1.val, n2.val, p_carry)
                n1 = n1.next
                n2 = n2.next

            # print("Carry: {}, Ones: {}".format(carry, ones))

            if result == None:
                result = ListNode(ones, None)
                r1 = result
            else:
                r1.next = ListNode(ones, None)
                r1 = r1.next
                p_carry = 0
            
            p_carry = carry

        if p_carry != 0:
            r1.next = ListNode(p_carry, None)

        return result

def add(d1, d2, pc):
    result = d1 + d2 + pc
    return (result // 10), (result % 10)

def print_list(n_list):

    print("Number Reversed:")

    node = n_list
    while not(node is None):
        print(node.val)
        node = node.next

    print()

def to_linked_list(n_list):
    root = None
    r = None
    for i in n_list:
        if root == None:
            root = ListNode(i, None)
            r = root
            continue

        r.next = ListNode(i, None)
        r = r.next

    return root

if __name__ == "__main__":

    l1 = [9,9,9,9,9,9,9]
    l2 = [9,9,9,9]

    ll1 = to_linked_list(l1)
    ll2 = to_linked_list(l2)

    sol = Solution()

    r1 = sol.addTwoNumbers(ll1, ll2)

    print_list(r1)